### PR TITLE
[FrameworkBundle][Serializer] Deprecate annotations

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | -
| License       | MIT
| Doc PR        | -

Part of https://github.com/symfony/symfony/issues/51381

### CODE REVIEW

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 1]
```diff
@@ -243,7 +252,7 @@ private function isKnownAttribute(string $attributeName): bool
     /**
      * @return object[]
      */
-    private function getClassAnnotations(\ReflectionClass $reflector): array
+    private function getClassAttributes(\ReflectionClass $reflector): array
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Please revert. This method reads Doctrine annotations only. Thus, its name is accurate. On 7.0, the method has been removed already.

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 2]
```diff
@@ -258,7 +267,7 @@ private function getClassAnnotations(\ReflectionClass $reflector): array
     /**
      * @return object[]
      */
-    private function getMethodAnnotations(\ReflectionMethod $reflector): array
+    private function getMethodAttributes(\ReflectionMethod $reflector): array
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Same

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 3]
```diff
@@ -273,7 +282,7 @@ private function getMethodAnnotations(\ReflectionMethod $reflector): array
     /**
      * @return object[]
      */
-    private function getPropertyAnnotations(\ReflectionProperty $reflector): array
+    private function getPropertyAttributes(\ReflectionProperty $reflector): array
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Same

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 4]
```diff
@@ -176,7 +175,7 @@ public function loadClassMetadata(ClassMetadataInterface $classMetadata): bool
         return $loaded;
     }
 
-    public function loadAnnotations(\ReflectionMethod|\ReflectionClass|\ReflectionProperty $reflector): iterable
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:190]: I wonder why this method is public anyway. All calls come from within this class. Let's just deprecate it without replacement. The new `loadAttributes()` method can be private.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php [hunk 5]
```diff
@@ -203,16 +202,26 @@ public function loadAnnotations(\ReflectionMethod|\ReflectionClass|\ReflectionPr
...
+            yield from $this->getPropertyAttributes($reflector);
         }
     }
 
+    /**
+     * @deprecated since Symfony 6.4, use "loadAttributes()" instead
+     */
+    public function loadAnnotations(\ReflectionMethod|\ReflectionClass|\ReflectionProperty $reflector): iterable
+    {
+        trigger_deprecation('symfony/serializer', '6.4', 'Method "%s()" is deprecated, use "loadAttributes()" instead.', __METHOD__);
```
Comment (mtarld) [inline src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: Isn't it triggering the deprecation twice? One with the phpdoc and one with the `trigger_deprecation`

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 1 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I think I've confused it with the Validator's AnnotationLoader, which reads attributes. I reverted

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 4 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php:190]: Changed visibility to private and removed the lines from the Changelog about it

File: src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php (hunk 5 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/Serializer/Mapping/Loader/AnnotationLoader.php]: I don't think so :thinking: If I remember correctly, the PHPDoc one is more for the IDE and DX than anything else

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -1935,12 +1935,7 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
                 $container->removeDefinition('serializer.mapping.cache_class_metadata_factory');
             }
 
-            $annotationLoader = new Definition(
-                AnnotationLoader::class,
-                [new Reference('annotation_reader', ContainerInterface::NULL_ON_INVALID_REFERENCE)]
-            );
-
-            $serializerLoaders[] = $annotationLoader;
+            $serializerLoaders[] = new Definition(AttributeLoader::class);
```
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: If you don't pass the annotation reader, annotations will stop working immediately. This is not what we want for 6.4, is it?

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (alexandre-daubois) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: That's right, I updated this part

**Comment by alexandre-daubois** [comment]
Rebased and conflicts fixed

**Comment by nicolas-grekas** [comment]
Thank you @alexandre-daubois.

**Pull request merged**